rpc module: refactor subscriptions to return `impl IntoSubscriptionResponse`

niklasad1 opened the pull request:
The goal with this PR is to simplify the subscription API from the server-side of things.

Previously the subscriptions were changed to return `Result` for ergonomics but the errors were never used and only logged after the latest refactoring. 

One problem with the current API is that "result error bailing" is hard to get right and might cause the error not to sent the client.

This PR adds a new trait `IntoSubscriptionCloseResponse` which can be implemented for customized types and it is implemented for `Result<(), StringError> and ()` by jsonrpsee

### Result

Once an error occurs it's sent out a subscription error notification

```rust
    module
        .register_subscription("my_sub", "my_sub", "my_unsub", |_, pending, _| async move {
            let mut sink = pending.accept().await?;
            sink.send("foo").await?;
            Ok(())
        })
        .unwrap();
```

The outcome is that the proc macro APIs can't assume that the return type will be `Result<(), Error>` anymore instead it could be any type that implement `IntoSubscriptionResponse`

niklasad1 committed:
refactor subscription to return `Option<Result<>>`

niklasad1 commented on core/src/server/rpc_module.rs:
```diff
@@ -838,14 +866,28 @@ impl<Context: Send + Sync + 'static> RpcModule<Context> {
-						if let Err(SubscriptionCallbackError::Some(msg)) = sub_fut.await {
-							tracing::warn!("Subscribe call `{subscribe_method_name}` failed: {msg}");
+						match sub_fut.await {
+							Some(Err(msg)) if tx.is_closed() => {
```
disable to send out notifications on subscription that hasn't been accepted.

jsdw commented on tests/tests/rpc_module.rs:
```diff
@@ -424,10 +419,8 @@ async fn reject_works() {
-			assert!(matches!(res, Ok(())));
-
-			Ok(())
+			pending.reject(err.into_owned()).await;
+			None
```
If you return an error instead of calling `pending.reject`, what will happen? And if you return an error as well as calling it, what happens?

I guess I'd hope that `pending.reject()` can actually go away in favour of returning an error, so that there's exactly one way to return an error from a subscription which also guarantees it's ended (the callback has returned).

niklasad1 replied:
it will deny the subscription will a default error and the returned error will be ignored as it's not valid to send out notifications before the subscription call has been answered.

niklasad1 replied:
hmm, good point we could actually take the error and call reject on the hood.

that's a nice API improvement IMHO

niklasad1 replied:
it's a bit tricky to implement with the middleware but I'll like to tackle it in another PR for sure.

niklasad1 committed:
address grumbles: change to Result<(), Option<M>>

This changes the subscription API again to return `Result<(), Option<SubscriptionMessage>>`
to work a little smoother with the combinators and it introduces an extention trait
to make convert it `Result<T, Error>` to `Result<T, Option<SubscriptionMessage>>`

This trait is implemented for types where it is possible and for custom types and tricky situations the
user has to implement themselves.

For instance it's not possible to implement on `TrySendError` because that depends on the use-case
i.e. whether the channel is full should be regarded as an error or not.

niklasad1 committed:
remove weird default errors

niklasad1 committed:
fix tests

niklasad1 commented on core/src/server/mod.rs:
```diff
@@ -32,3 +32,48 @@ pub mod helpers;
+///
+/// This is implemented for the types in jsonrpsee where the behavior is not application dependent.
+/// For other types you have to implement this trait or deal with it manually.
+pub trait MapSubscriptionError<T> {
```
I'm not sure that we should have this trait, it works good for user-defined types but for other types from jsonrpsee such as `TrySendError`, `SendTimeoutError` then folks has to move them into a new type to not bump orphan rule issues.

niklasad1 replied:
also to `map_err` works now so perhaps better just to have that

jsdw replied:
Mmm I'm not super keen on this trait; if I was going for any trait it'd be something like:

```rust
trait IntoResponse {
    fn into_response(self) -> SubscriptionCloseResponse
}

impl IntoResponse for Option<T> where T: Into<SubscriptionMessage> {
    fn into_response(self) ->  SubscriptionCloseResponse {
         // None means nothing returned, Some(msg) means a non-error msg is returned
    }
}

impl IntoResponse for Result<T, E> {
    fn into_response(self) ->  SubscriptionCloseResponse {
          // Err(e) means error returned, Ok(m) means a non-error msg is returned
    }
}
```

But should `Ok(())` and `None` mean that no message is returned at all, or with `Ok(())` would you get back `[]` as a message? It feels like it would be safer if no message came back (otherwise using `Result` there is no way _not_ to return some message if all goes ok), and you just have to explicitly send one yourself (or register one to be sent on the sink).

So I guess I'd only support `Result<(),IntoError>` and `Option<()>`, and then we can ergonomically return errors using `Result` and `?` or ergonomically return but emit no errors with `Option` and `None`.

So one could write:

```rust
module.register_subscription("my_sub", "my_sub", "my_unsub", |_, pending, _| async move {
	// Do stuff; can use ? to return errors here
    pending.accept().await?;

	Ok(())
})
```

Or one can opt to not return errors like this:

```rust
module.register_subscription("my_sub", "my_sub", "my_unsub", |_, pending, _| async move {
	// Due to return type, need to `.ok()` explicitly to throw away errors.
    pending.accept().await.ok()?;

	None
})
```

And maybe if you want to send a default closing message some api like:

```rust
module.register_subscription("my_sub", "my_sub", "my_unsub", |_, pending, _| async move {
    let mut sink = pending.accept().await.ok()?;

    // register message to be sent when closes (presumably if an error is returned because we
    // allow this via `Result`, we will either send this just prior to the error, or not send it at all 
    // (don't know that it matters which one!):
    sink.send_on_close(|| SomeMessage);

	None
})
```

But if all of that is sortof annoying or too much effort, then the alternative which I also think is perfectly reasonable is to only return `()` (and force error handling to be done by the user), or only return `Result<(), SomeError>` to force the user to do `pending.accept().await? else { /* handle error manually each time */ }` but make it easy if the user wants to return errors by default.

Really, the only thing I don't like about the currently released interface is that you can return the error and then it's silently discarded, which I think may be surprising.

jsdw replied:
I guess the current API you've settled on without the trait is pretty much equivalent to returning `()` in terms of usability, except for allowing the user to return some message (or error I think?) if they like. So I'm also happy with that offhand if that's what you prefer!

niklasad1 replied:
The reason why I didn't go down the `into_response` route is because it didn't work with `?` but maybe I did something wrong when I tried it.

niklasad1 replied:
so yeah, I got it working as well but as the type inference forces the `async block` to return `Result or Option` with the design which makes it not possible to ignore some error but in another scenario return an error if that makes sense :)

[example here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d7429031e971510c4ffd00f347df433b)

jsdw replied:
> which makes it not possible to ignore some error but in another scenario return an error in the same async block

I think you mean that you want some errors to be ignored (eg connection closed) when `?` is applied on them, while letting other errors propagate to the user?

`IntoResponse` could always expect an error enum to be returned and then "connection closed" style errors can be manually filtered out if they are seen? Or errors can impl `Into<ResponseError>` which would have the error info and also note whether it should be sent to the user or not (http libraries sometimes do similar stuff; sending 4xx errors to users and only logging 5xx errors on the server while sending the user some generic message)

But I might be missing the point so apologies if so!

niklasad1 replied:
Yeah, that's true maybe I'm a bit narrow minded but initially the idea was that `return Option == ignore error` and `return Err == always propagate the error` to make it explicit for the user.

Having what you suggested is nice but isn't really what I had in mind and easy to "bail early with ?`....

jsdw replied:
> but initially the idea was that return Option == ignore error and return Err == always propagate the error to make it explicit for the user.

Yup that was definitely the thinking but I thought you were raising that not all errors should be returned by default, but I may have misunderstood!

niklasad1 replied:
yeah sorry I was more picking hair on the that you can't really return `Result and Option` from the substration callback and tried to elaborate that it's ok but yeah we could ignore some error but it doesn't really matter if the connection is closed anyway :P

I think I prefer your suggestion anyway, I'll give it a try again

niklasad1 committed:
revert weird trait

niklasad1 committed:
Merge remote-tracking branch 'origin/master' into na-subscription-api-option-result

niklasad1 committed:
Merge remote-tracking branch 'origin/master' into na-subscription-api-option-result

niklasad1 committed:
introduce `IntoSubscriptionResult`

niklasad1 committed:
refactor: introduce subscription mod

niklasad1 committed:
impl StdError for subscription error types

niklasad1 committed:
fix nits + tests

niklasad1 committed:
fix doc tests

niklasad1 committed:
remove deadcode

niklasad1 commented on server/src/tests/ws.rs:
```diff
@@ -687,9 +695,9 @@ async fn ws_server_backpressure_works() {
-				let sink = pending.accept().await?;
-				let n = SubscriptionMessage::from_json(&1).unwrap();
-				let bp = SubscriptionMessage::from_json(&2).unwrap();
+				let sink = pending.accept().await.map_err(|e| anyhow::anyhow!("{:?}", e))?;
```
```suggestion
				let sink = pending.accept().await?;
```

niklasad1 committed:
Update server/src/tests/ws.rs

jsdw commented on proc-macros/src/lib.rs:
```diff
@@ -293,23 +293,21 @@ pub(crate) mod visitor;
 ///
-///             let msg1 = SubscriptionMessage::from_json(&"Response_A")?;
-///             let msg2 = SubscriptionMessage::from_json(&"Response_B")?;
+///             let msg1 = SubscriptionMessage::from_json(&"Response_A").expect("Infallible");
```
I don't get this yet; wouldn't `from_json` reject the invalid json "Response_A"?

jsdw commented on proc-macros/src/lib.rs:
```diff
@@ -293,23 +293,21 @@ pub(crate) mod visitor;
-///             sink.send(msg1).await?;
-///             sink.send(msg2).await?;
+///             sink.send(msg1).await.map_err(|_| None)?;
+///             sink.send(msg2).await.map_err(|_| None)?;
```
Since we're just throwing away the message, does it make sense for the example to `map_err(|_| None)` everywhere insetad of returning some `anyhow::Error` or something?

niklasad1 replied:
I haven't updated this file yet :)

jsdw commented on benches/helpers.rs:
```diff
@@ -151,11 +151,11 @@ pub async fn ws_server(handle: tokio::runtime::Handle) -> (String, jsonrpsee::se
 			|_params, pending, _ctx| async move {
-				let sink = pending.accept().await?;
+				let sink = pending.accept().await.unwrap();
 				let msg = SubscriptionMessage::from("Hello");
-				sink.send(msg).await?;
+				sink.send(msg).await.unwrap();
```
Nit; is there any harm in using `?` here?

niklasad1 replied:
no :)

I didn't change it after playing around with different approaches.

jsdw commented on core/src/server/error.rs:
```diff
@@ -0,0 +1,96 @@
+
+/// The error returned while accepting or rejecting a subscription.
+#[derive(Debug, Copy, Clone, thiserror::Error)]
+pub struct SubscriptionAcceptRejectError;
```
Nit: `SubscriptionAcceptRejectError` is a bit confusing to me; something like `PendingSubscriptionAcceptError` maybe? (it's an error accepting a pending subscription)

jsdw commented on core/src/server/error.rs:
```diff
@@ -0,0 +1,96 @@
+#[derive(Debug, Copy, Clone, thiserror::Error)]
+pub struct SubscriptionAcceptRejectError;
+
+impl std::fmt::Display for SubscriptionAcceptRejectError {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		f.write_str("The remote peer closed the connection")
+	}
+}
```
Since `thiserror` is being used, I think you could also add an `#[error("The remote peer closed the connection")]` to the type for the same effect (and for DisconnectError)

niklasad1 replied:
nice catch, didn't know that it worked on structs

jsdw commented on core/src/server/rpc_module.rs:
```diff
@@ -721,39 +561,73 @@ impl<Context: Send + Sync + 'static> RpcModule<Context> {
+	///
+	/// This is implemented for `Result<T, E>` and `Option<T>`.
+	///
+	/// It's recommended to use `Result` if you want to propagate the error as special error notification
```
I haven't checked yet, but if the subscription returns an error before `pending.accept()` does that also come back (but as an error to the subscription request rather than an error notification). Maybe the text should be more general and just say something like "Use `Result` if you want the error to be sent back to the user. Use `Option` if you'd like to avoid sending any errors back to the user."?

jsdw replied:
Ah nevermind; it's a different thing and the user has to `reject()` with an error, which makes sense, so ignore this :)

niklasad1 replied:
This applies only to subscription notifications i.e, subscription messages and before that the subscription has been accepted we don't allow any messages to be sent. So those are ignored unless the `reject` is called explicitly. 

However, it would be nice to incorporate that in the callback and let remove `reject API`

TLDR: Returning stuff before `accept` has no effect.

niklasad1 replied:
So, I want to keep it out of this PR but indeed this is hairy stuff to understand.

jsdw replied:
Yeah makes sense; I can imagine that if the user returns an IntoResponse thing before calling `pending.accept()`, that maybe it could return the error they gave back (if a `Result` type was used) or just return some default error. And then maybe keep `pending.reject(explicit_error)` around so that even if the user wants to return `Option` from the subscription, they can still send a non-default error back here if they want.

But yeah, definitely can be a separate PR if wanted :)

jsdw commented on core/src/server/rpc_module.rs:
```diff
@@ -721,39 +561,73 @@ impl<Context: Send + Sync + 'static> RpcModule<Context> {
-	/// `SubscriptionSink::send` before returning from the async block.
+	/// Another option is to implement [`crate::server::IntoSubscriptionResponse`] if you want customized behaviour.
+	///
+	/// The error notification has the following format:
```
If I'm right above, maybe also add a note on what the error looks like if the Subscription wasn't accepted yet?

jsdw replied:
Ah I see; it is a different thing!

jsdw commented on core/src/server/rpc_module.rs:
```diff
@@ -721,39 +561,73 @@ impl<Context: Send + Sync + 'static> RpcModule<Context> {
+	///     let x = match params.one::<usize>() {
+	///         Ok(x) => x,
+	///         Err(e) => {
+	///            pending.reject(ErrorObjectOwned::from(e)).await;
```
I wondered whether a user could return an error for this too, but realised that you _have_ to return some sort of error for this, so yeah I see why it's separate and it makes sense to me!

(I guess 99% of the time people will just `pending.accept()` anyway!)

jsdw commented on core/src/server/subscription.rs:
```diff
@@ -0,0 +1,475 @@
+	Err(SubscriptionMessage),
+}
+
+impl<T> IntoSubscriptionResponse for Option<T>
```
Nit: I feel like this name, `IntoSubscriptionResponse`, should be consistent with the name of the thing that's it converts stuff into (ie maybe `SubscriptionResponse` rather than `SubscriptionCloseResponse`)

jsdw commented on core/src/server/subscription.rs:
```diff
@@ -0,0 +1,475 @@
+}
+
+#[derive(Debug, Clone, Copy)]
+pub(crate) enum SubNotifResultOrError {
+	Result,
+	Error,
+}
```
Nit: since `Result` is a thing with an `Ok` or `Err` variant already, maybe `Response, Error` makes more sense here than `Result, Error` for the variant names?

jsdw commented on core/src/server/subscription.rs:
```diff
@@ -0,0 +1,475 @@
+#[must_use = "PendningSubscriptionSink does nothing unless `accept` or `reject` is called"]
+pub struct PendingSubscriptionSink {
+	/// Sink.
+	pub(crate) inner: MethodSink,
+	/// MethodCallback.
+	pub(crate) method: &'static str,
+	/// Shared Mutex of subscriptions for this method.
+	pub(crate) subscribers: Subscribers,
+	/// Unique subscription.
+	pub(crate) uniq_sub: SubscriptionKey,
+	/// ID of the `subscription call` (i.e. not the same as subscription id) which is used
+	/// to reply to subscription method call and must only be used once.
+	pub(crate) id: Id<'static>,
+	/// Sender to answer the subscribe call.
+	pub(crate) subscribe: mpsc::Sender<MethodResponse>,
+	/// Subscription permit.
+	pub(crate) permit: SubscriptionPermit,
```
Nit: I'm wondering how many of these need to be `pub(crate)` and whether we can hide a bit more

niklasad1 replied:
it's instantiated from the `RpcModule` so needs to be `pub(crate)` or having `pub(crate) fn new`

jsdw commented on core/src/server/subscription.rs:
```diff
@@ -0,0 +1,475 @@
+
+	/// Attempt to accept the subscription and respond the subscription method call.
+	///
+	/// Returns `None` if the connection is already closed, `Some(SubscriptionSink)` otherwise.
```
Comment out of date

jsdw commented on server/src/tests/ws.rs:
```diff
@@ -404,11 +404,13 @@ async fn register_methods_works() {
 	assert!(module
-		.register_subscription("subscribe_hello", "subscribe_hello", "unsubscribe_hello", |_, _, _| async { Ok(()) })
+		.register_subscription("subscribe_hello", "subscribe_hello", "unsubscribe_hello", |_, _, _| async {
+			Ok::<_, String>(())
```
I guess these things could also be `Some(())` now to avoid type inference on error stuff?

niklasad1 committed:
cleanup to make review easier

niklasad1 committed:
SubscriptionAcceptError -> PendingSubscriptionAcceptError

niklasad1 commented on core/src/server/rpc_module.rs:
```diff
@@ -474,7 +316,7 @@ impl Methods {
 			Some(MethodCallback::Subscription(cb)) => {
 				let conn_state =
-					ConnState { conn_id: 0, id_provider: &RandomIntegerIdProvider, subscription_permit: p1 };
+					SubscriptionState { conn_id: 0, id_provider: &RandomIntegerIdProvider, subscription_permit: p1 };
```
it's just to `mock` a `SubscriptionPermit` when it's used for tests/raw calls on RPC module without a "real transport".

niklasad1 replied:
hopefully more readable in https://github.com/paritytech/jsonrpsee/pull/1034/commits/438371b99873f691cd848ceb0b307cc68dbcb905

niklasad1 committed:
make it clear that SubscriptionPermit is mocked

niklasad1 committed:
remove deadcode

niklasad1 committed:
Merge remote-tracking branch 'origin/master' into na-subscription-api-option-result

niklasad1 committed:
fix doc links

niklasad1 commented on core/src/server/helpers.rs:
```diff
@@ -191,59 +189,6 @@ pub fn prepare_error(data: &[u8]) -> (Id<'_>, ErrorCode) {
-/// Wrapper over [`tokio::sync::Notify`] with bounds check.
-#[derive(Debug, Clone)]
-pub struct BoundedSubscriptions {
-	resource: Arc<Notify>,
```
these `Arc<Notify>` are not used anymore

niklasad1 committed:
make register_subscription more readable

niklasad1 committed:
remove Option impl for IntoSubscriptionCloseResponse

niklasad1 committed:
fix tests

niklasad1 committed:
add test for subscription return type

niklasad1 committed:
fix nits

jsdw commented on proc-macros/tests/ui/incorrect/sub/sub_unsupported_field.rs:
```diff
@@ -4,7 +4,7 @@ use jsonrpsee::proc_macros::rpc;
 pub trait UnsupportedField {
 	#[subscription(name = "sub", unsubscribe = "unsub", item = u8, magic = true)]
-	async fn sub(&self);
+	async fn sub(&self) -> RpcResult<()>;
```
I guess you could `impl IntoSubscriptionResponse` for `()` too (ie no message or error sent back) if you want to allow empty return types.

niklasad1 replied:
yeah I was thinking the same, why not

niklasad1 replied:
implemented now

niklasad1 committed:
impl `IntoSubscriptionCloseResponse` for ()

niklasad1 committed:
remove useless ui test

niklasad1 commented on core/src/server/error.rs:
```diff
@@ -0,0 +1,91 @@
+// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::server::subscription::SubscriptionMessage;
```
```suggestion
use crate::server::SubscriptionMessage;
```

niklasad1 committed:
Update core/src/server/error.rs

niklasad1 commented on core/src/server/rpc_module.rs:
```diff
@@ -495,7 +336,7 @@ impl Methods {
 		tracing::trace!("[Methods::inner_call] Method: {}, response: {:?}", req.method, response);
 
-		(response, rx, p2)
+		(response, rx)
```
this permit was never used

niklasad1 committed:
register_subscription: change order of parameters

niklasad1 commented on core/src/server/error.rs:
```diff
@@ -0,0 +1,91 @@
+	Closed(SubscriptionMessage),
+}
+
+/// The error returned while accepting or rejecting a subscription.
```
```suggestion
/// The error returned while accepting a subscription.
```

niklasad1 committed:
Update core/src/server/error.rs

niklasad1 committed:
Merge remote-tracking branch 'origin/na-subscription-api-option-result' into na-subscription-api-option-result

niklasad1 commented on core/src/server/subscription.rs:
subscription related types moved to this file

niklasad1 commented on jsonrpsee/src/lib.rs:
```diff
@@ -90,7 +90,7 @@ cfg_types! {
 
 cfg_server! {
-	pub use jsonrpsee_core::server::rpc_module::{RpcModule, SubscriptionSink, PendingSubscriptionSink, SubscriptionMessage, DisconnectError, TrySendError, SendTimeoutError};
+	pub use jsonrpsee_core::server::*;
```
this is already re-exported by the server maybe not needed

niklasad1 commented on proc-macros/src/lib.rs:
```diff
@@ -292,27 +314,40 @@ pub(crate) mod visitor;
 ///         }
+///
+///         // If one doesn't want sent out a close message when a subscription terminates or treat
+///         // error as subscription error notifications then it's possible to implement
```
```suggestion
///         // errors as subscription error notifications then it's possible to implement
```

niklasad1 committed:
Update proc-macros/src/lib.rs

niklasad1 committed:
fix doc link

niklasad1 committed:
Merge remote-tracking branch 'origin/na-subscription-api-option-result' into na-subscription-api-option-result

niklasad1 committed:
proc macro: compile test for !Result

niklasad1 committed:
hacky solution

niklasad1 committed:
add niklas ugly hack for providing default value

niklasad1 committed:
fix hacks

niklasad1 committed:
cargo fmt

jsdw commented on core/src/server/subscription.rs:
```diff
@@ -0,0 +1,444 @@
+		match self {
+			Ok(msg) => match SubscriptionMessage::from_json(&msg) {
+				Ok(m) => SubscriptionCloseResponse::Some(m),
+				Err(e) => SubscriptionCloseResponse::Err(e.to_string().into()),
```
Just checking; is this Err from attempting to serialize the message? I wonder whether we should send nothing back to the user and just log it in that case, since it's more of an internal error?

niklasad1 replied:
yeah, that error is useless anyway I think `serialization failed` :P

jsdw commented on core/src/server/subscription.rs:
```diff
@@ -0,0 +1,444 @@
+
+impl IntoSubscriptionCloseResponse for SubscriptionCloseResponse {
+	fn into_response(self) -> SubscriptionCloseResponse {
+		SubscriptionCloseResponse::None
```
Shouldn't this return `self`?

jsdw replied:
Maybe the change isn't on github yet; just want to check that this is not missed :)

jsdw commented on core/src/server/error.rs:
```diff
@@ -0,0 +1,91 @@
+#[error("The remote peer closed the connection")]
+pub struct PendingSubscriptionAcceptError;
+
+impl std::fmt::Display for DisconnectError {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		f.write_str("Closed")
+	}
+}
```
Could add `#[error("Closed")]` to DisconnectError instead?

jsdw commented on core/src/server/subscription.rs:
```diff
@@ -0,0 +1,444 @@
+	}
+}
+
+/// Wrapper [`tokio::sync::Semaphore`] for subscriptions.
```
```suggestion
/// Wrap [`tokio::sync::Semaphore`] for subscriptions.
```

jsdw commented on examples/examples/ws_pubsub_broadcast.rs:
```diff
@@ -116,20 +112,19 @@ async fn pipe_from_stream_with_bounded_buffer(
-			Either::Right((_, _)) => {
-				break;
-			}
+			// Send back back the error.
+			Either::Right((Some(Err(e)), _)) => break Err(e.into()),
```
It's nice that it's easy to return an error like this now!

niklasad1 replied:
yeah, the only annoyance is if you want a JSON value here.

Then you would need to `serde_json::to_string(&v)` and the `Subscription::from(String)` would add an extra `""kakak""`.... but that's sacrificed with the StdError/ToString here :(

jsdw reviewed (commented):
Looks awesome; just a couple of small comments/questions :)

niklasad1 commented on server/src/tests/ws.rs:
```diff
@@ -421,8 +431,12 @@ async fn register_methods_works() {
 	assert!(matches!(
-		module
-			.register_subscription("subscribe_hello", "subscribe_hello", "subscribe_hello", |_, _, _| async { Ok(()) }),
+		module.register_subscription::<SubscriptionResult, _, _>(
+			"subscribe_hello",
+			"subscribe_hello",
+			"subscribe_hello",
+			|_, _, _| async { Ok(()) }
+		),
```
perhaps your idea is slightly better but as long we don't have `Ok::<_, Error>::(())` then I'm happy

niklasad1 committed:
Update core/src/server/subscription.rs

Co-authored-by: James Wilson <james@jsdw.me>

niklasad1 commented on core/src/server/subscription.rs:
```diff
@@ -0,0 +1,444 @@
+#[must_use = "PendningSubscriptionSink does nothing unless `accept` or `reject` is called"]
+pub struct PendingSubscriptionSink {
+	/// Sink.
+	pub(crate) inner: MethodSink,
+	/// MethodCallback.
+	pub(crate) method: &'static str,
+	/// Shared Mutex of subscriptions for this method.
+	pub(crate) subscribers: Subscribers,
+	/// Unique subscription.
+	pub(crate) uniq_sub: SubscriptionKey,
+	/// ID of the `subscription call` (i.e. not the same as subscription id) which is used
+	/// to reply to subscription method call and must only be used once.
+	pub(crate) id: Id<'static>,
+	/// Sender to answer the subscribe call.
+	pub(crate) subscribe: mpsc::Sender<MethodResponse>,
+	/// Subscription permit.
+	pub(crate) permit: OwnedSemaphorePermit,
```
I could make it `pub(crate) new -> PendingSubscriptionSink` but I like this better with named initialization

niklasad1 replied:
right, we could limit that to the `rpc_module` to be even more strict

niklasad1 committed:
fix grumbles

niklasad1 committed:
Merge remote-tracking branch 'origin/na-subscription-api-option-result' into na-subscription-api-option-result

niklasad1 commented on core/src/server/subscription.rs:
```diff
@@ -0,0 +1,444 @@
+		let response =
+			MethodResponse::response(self.id, &self.uniq_sub.sub_id, self.inner.max_response_size() as usize);
+		let success = response.success;
+		self.inner.send(response.result.clone()).await.map_err(|_| PendingSubscriptionAcceptError)?;
+		self.subscribe.send(response).await.map_err(|_| PendingSubscriptionAcceptError)?;
```
Opened an issue for it and added some inline comments, hopefully more understandable now

niklasad1 committed:
Merge remote-tracking branch 'origin/master' into na-subscription-api-option-result

niklasad1 commented on core/src/lib.rs:
```diff
@@ -58,16 +58,11 @@ cfg_client! {
-/// The return type of the subscription's method for the rpc server implementation.
-///
-/// **Note**: The error does not contain any data and is discarded on drop.
-pub type SubscriptionResult = Result<(), SubscriptionCallbackError>;
```
So, I removed this because folks have to convert their subscriptions it may be possible that they sent out subscription close errors if they doesn't read the docs carefully enough and just keep `Result<T, E>` in the subscriptions.

we could keep it as well.

niklasad1 commented on core/src/server/subscription.rs:
```diff
@@ -0,0 +1,486 @@
+	NotifErr(SubscriptionMessage),
+}
+
+impl<T, E> IntoSubscriptionCloseResponse for Result<T, E>
```
I think we `()` will be serialized as `null` which is annoying

niklasad1 replied:
we could also implement that for `Result<(), Error>` instead

jsdw replied:
As we said in the call, that sounds like a good idea to me!

I even wonder whether it makes sense for `Error` to be a concrete type (`?` will ".into()" other error types into it if possible anyway). This would remove the need to specify any generic types anywhere :)

niklasad1 replied:
What I had in mind with this is that you can also inject your own JSON here such as

```
result: { "my": 120, "hehe": { "1":1} }
```

if we put `Error` there that is not possible it would just be:

```
result: "<err.to_string()>"
```

however, `ToString` here doesn't work super well either I think we would add extra string quotes ("")

niklasad1 replied:
sorry, I didn't realize that I wrote `Error` above I meant `E: ToString` :)

jsdw replied:
> What I had in mind with this is that you can also inject your own JSON here such as
> 
> ```
> result: { "my": 120, "hehe": { "1":1} }
> ```
> 
> if we put `Error` there that is not possible it would just be:
> 
> ```
> result: "<err.to_string()>"
> ```
> 
> however, `ToString` here doesn't work super well either I think we would add extra string quotes ("") such if one does:
> 
> ```
> sink.send().await.map_err(|_| serde_json::to_string(&MyError).unwrap())?;
> ```
> 
> lemme create a test for that

`E` impl's `ToString` at the moment, but for instance if we had a concrete `StringError` type for instance that serialized to a string correctly, and add `impl <T: ToString> From<T> for StringError` then all `?`'s would continue to work (as long as the errors impl Display which they generally always do) and no generics would be needed.

Dunno; maybe that doesn't make sense?

niklasad1 replied:
ok, it won't work with the proc macro code basically it's not possible to have the assumption that `?` will work it could be `()` or some other type.

Thus, I could just think of something like:

```rust
  if ret_ty == "Result" {
                #into_sub_response::into_response(context.as_ref().#rust_method_name(subscription_sink, #params_seq).await.map_err(Into::into))
  } else {
    #into_sub_response::into_response(context.as_ref().#rust_method_name(subscription_sink, #params_seq).await
  }
```

niklasad1 replied:
Implemented something similar as we have for the client to check the return type ident but I'm not super happy with it.

It works but a little hacky but probably fine here.

jsdw replied:
> ok, it won't work with the proc macro code basically it's not possible to have the assumption that ? will work it could be () or some other type.

Ah I didn't think it would impact the proc macro; I'd assuemd that any IntoSubscriptionCloseResponse thing would "just work" as a return type so it didn't matter. Just happens that for user facing code if they want to return a Result type then they can use "?"'s and it'll work out?

niklasad1 replied:
fixed now

niklasad1 committed:
fix tests

niklasad1 committed:
add test for null notification

niklasad1 committed:
impl IntoSubscriptionResponse for Result<(), E>

niklasad1 committed:
more descriptive error impls

niklasad1 commented on core/src/server/error.rs:
```diff
@@ -0,0 +1,85 @@
+#[derive(Debug, thiserror::Error)]
+pub enum TrySendError {
+	/// The channel is closed.
+	#[error("Closed")]
```
maybe I can get rid of those `SubscriptionMessage` there looks a bit weird and just convert them in the `SubscriptionSink impls` instead

lexnv commented on core/src/server/subscription.rs:
```diff
@@ -0,0 +1,475 @@
+		&self.sub_id
+	}
+
+	/// Returns `Some((val, sub_id))` for the next element of type T from the underlying stream,
```
nit: Do we need to update the comment here? This returns Some(Ok((val, sub_id)) and does not panic anymore. Also, we could remove the `let res` below and return directly, I guess that was used for debugging :D

niklasad1 replied:
no, it won't compile otherwise some lifetime issue without the let res binding ^^

jsdw commented on core/src/server/subscription.rs:
```diff
@@ -0,0 +1,475 @@
+
+impl IntoSubscriptionCloseResponse for SubscriptionCloseResponse {
+	fn into_response(self) -> Self {
+		Self::None
```
```suggestion
		self
```

Surely? :D

niklasad1 replied:
ah, nice catch

lexnv commented on proc-macros/tests/ui/incorrect/sub/sub_no_item.rs:
```diff
@@ -4,7 +4,7 @@ use jsonrpsee::proc_macros::rpc;
 pub trait NoSubItem {
 	#[subscription(name = "sub")]
-	async fn sub(&self);
+	async fn sub(&self) -> RpcResult<()>;
```
nit: Should we add a `sub_no_item.stderr` for this and the others missing the `.stderr`?

niklasad1 replied:
we already got one and I think this happens before the `RpcResult` is checked but let's revert these UI tests

it should work for both `Result and ()` anyway

niklasad1 committed:
Update core/src/server/subscription.rs

Co-authored-by: James Wilson <james@jsdw.me>

niklasad1 commented on proc-macros/src/render_server.rs:
```diff
@@ -76,11 +76,6 @@ impl RpcDescription {
-			// For ergonomic reasons, the server's subscription method should return `SubscriptionResult`.
-			let return_ty = self.jrps_server_item(quote! { core::SubscriptionResult });
-			let output: ReturnType = parse_quote! { -> #return_ty };
-			sub_sig.sig.output = output;
```
No, if anything we should check if the type returns `impl IntoSubscriptionResponse`  but that is so tricky/impossible so why even bother?

The compile error is not that bad:

```bash
error[E0277]: the trait bound `std::option::Option<usize>: IntoSubscriptionCloseResponse` is not satisfied
  --> examples/examples/proc_macro.rs:38:1
   |
38 | #[rpc(server, client, namespace = "state")]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `IntoSubscriptionCloseResponse` is not implemented for `std::option::Option<usize>`
   |
   = help: the following other types implement trait `IntoSubscriptionCloseResponse`:
             ()
             Result<(), E>
             SubscriptionCloseResponse
   = note: this error originates in the attribute macro `rpc` (in Nightly builds, run with -Z macro-backtrace for more info)

```

niklasad1 committed:
update docs: Subscription::next

niklasad1 committed:
revert UI tests

niklasad1 committed:
proc macro tests: add subscription -> ()

niklasad1 committed:
remove unused tokio feature

niklasad1 committed:
add `SerializeAsStr` error

niklasad1 committed:
show failing test

niklasad1 committed:
fix tests

niklasad1 committed:
revert JSON str parsing

niklasad1 committed:
show James

niklasad1 committed:
push hack to show James what I mean

lexnv commented on proc-macros/src/helpers.rs:
```diff
@@ -193,9 +208,52 @@ pub(crate) fn extract_doc_comments(attrs: &[syn::Attribute]) -> TokenStream2 {
+	if type_name.ident == "Result" {
+		// Result<T, E> should have 2 generic args.
+		if args.len() != 2 {
+			panic!("Result type aliases are not allowed only `std::result::Result<T, E>` and `jsonrpsee::core::RpcResult` are supported");
```
nit: Would it be possible to also include the `Span` of the return type in the error?

Maybe returing the `TokenStream` or even panicking with the span info 
```rust
return Err(syn::Error::new(ty.span(), "msg").to_compile_error());
```

We don't necessarily the return types path `std::result::Result<T, E>` maybe renaming to `Result<T, E>` or similar?

niklasad1 replied:
thanks, I'll revert his and return `Result<(), StirngError` instead I don't like this hacky code

niklasad1 replied:
it was more to show James what I meant

niklasad1 committed:
SubscriptionResult = Result<(), StringError>

niklasad1 commented on tests/tests/rpc_module.rs:
```diff
@@ -502,3 +495,36 @@ async fn bounded_subscription_works() {
 }
+
+#[tokio::test]
+async fn serialize_sub_error_adds_extra_string_quotes() {
```
@jsdw @lexnv 

This is a test to illustrate that `StringError` will be serialized with extra string quotes such as `"error":"{...}" even if the value is JSON

We could actually parse the string before adding the `"<string>"` but I came to the solution that is too opinionated and likely better for folks to implement their own type then.

lexnv replied:
Yep, that does make sense! In the end, we will be wrapping the message (may it be a string or already a valid json) with "{ msg }". I'm guessing this is not the common behavior and we don't need to provide a sort of wrapped error to ensure we don't add the extra quotes.
I do wonder, do we have in substrate subscriptions that return directly a valid json for the error? I'd expect that to be unlikely

niklasad1 committed:
fix tests

lexnv reviewed (approved):
LGTM! Amazing work here!

niklasad1 committed:
address grumbles

jsdw reviewed (approved):
Awesome stuff! :)

niklasad1 commented on core/src/server/subscription.rs:
```diff
@@ -0,0 +1,484 @@
+
+/// Type-alias for subscribers.
+pub type Subscribers = Arc<Mutex<FxHashMap<SubscriptionKey, (MethodSink, mpsc::Receiver<()>)>>>;
+/// Subscription permit
```
```suggestion
/// Subscription permit.
```

niklasad1 committed:
Update core/src/server/subscription.rs

niklasad1 commented on proc-macros/src/lib.rs:
```diff
@@ -86,7 +86,7 @@ pub(crate) mod visitor;
 ///
 ///     #[subscription(name = "subscribe", item = "String")]
-///     async fn sub(&self);
+///     async fn sub(&self) -> Result<(), Error>;
```
```suggestion
///     async fn sub(&self) -> SubscriptionResult;
```

niklasad1 committed:
Update proc-macros/src/lib.rs

niklasad1 commented on proc-macros/src/lib.rs:
```diff
@@ -99,8 +99,8 @@ pub(crate) mod visitor;
-///     // Note that `pending_subscription_sink` and `SubscriptionResult` were added automatically.
-///     async fn sub(&self, pending: PendingSubscriptionSink) -> SubscriptionResult;
+///     // Note that `pending_subscription_sink` was added automatically.
+///     async fn sub(&self, pending: PendingSubscriptionSink) -> Result<(), Error>;
```
```suggestion
///     async fn sub(&self, pending: PendingSubscriptionSink) -> SubscriptionResult;
```

niklasad1 committed:
Update proc-macros/src/lib.rs

niklasad1 committed:
remove unused error variant SubscriptionCallbackError
